apps/web-ext/src/components/windowMessaging/window-messaging.ts

```diff
@@ -0,0 +1,45 @@
+import { GetDataType } from "@webext-core/messaging";
+
+/** window & targetOrigin 指定可能な defineWindowMessaging() */
+export function defineWindowMessaging<T extends Record<string, unknown>>() {
+  function sendMessage<TType extends keyof T>(
+    type: TType,
+    data: GetDataType<T[TType]>,
+    window?: Window | null,
+    targetOrigin?: URL["origin"],
+  ): void {
+    if (targetOrigin) {
+      window?.postMessage({ type, data }, targetOrigin);
+    } else {
+      window?.postMessage({ type, data });
+    }
+  }
+  function onMessage<TType extends keyof T>(
+    type: TType,
+    handler: (event: MessageEvent<GetDataType<T[TType]>>) => void,
+  ): () => void {
+    function listener(
+      event: MessageEvent<{
+        type: keyof T;
+        data: GetDataType<T[TType]>;
+      }>,
+    ) {
+      if (!(event.isTrusted && event.data.type === type)) return;
```

Comment:
もともとのコードからNAND結合なのであまり重要でもない気にしなくていいポイントかもしれないのですが、OR結合のほうがほんの少し認知負荷低めな印象を受けました:

```suggestion
      if (!event.isTrusted || event.data.type !== type) return;

```

あるいは

```ts
if (!event.isTrusted) return;
if (event.data.type !== type) return;
```